File: python/src/br_validator/ie/_pe.py
"""PE: verified against the official SEFAZ-PE "Roteiro de Crítica da
Inscrição Estadual" for the e-Fisco system
(sintegra.gov.br/Cad_Estados/cad_PE.html), including its fully worked
example (0321418-40). Format: 7 digits + 2 check digits.
"""

import re

from .._shared import remove_non_digits
from ._mod11 import mod11_check_digit, weighted_sum

_ALLOWED_CHARS = re.compile(r"^[\d\-\s]+$")
_FIRST_DIGIT_WEIGHTS = [8, 7, 6, 5, 4, 3, 2]
_SECOND_DIGIT_WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2]


def normalize(value: str) -> str:
    return remove_non_digits(value)


def is_valid(value: str) -> bool:
    if not _ALLOWED_CHARS.match(value):
        return False

    v = normalize(value)

    if len(v) != 9:
        return False

    base = v[:7]
    first_digit = mod11_check_digit(weighted_sum(base, _FIRST_DIGIT_WEIGHTS))

    if first_digit != int(v[7]):
        return False

    second_digit = mod11_check_digit(weighted_sum(base + str(first_digit), _SECOND_DIGIT_WEIGHTS))

    return second_digit == int(v[8])


def format(value: str) -> str:
    v = normalize(value)

    if len(v) != 9:
        return value

    return f"{v[:7]}-{v[7:9]}"

File: ruby/lib/br_validator/ie/pe.rb
# frozen_string_literal: true

require_relative "../shared"
require_relative "mod11"

module BrValidator
  module Ie
    # PE: verified against the official SEFAZ-PE "Roteiro de Crítica da
    # Inscrição Estadual" for the e-Fisco system
    # (sintegra.gov.br/Cad_Estados/cad_PE.html), including its fully
    # worked example (0321418-40). Format: 7 digits + 2 check digits.
    module Pe
      module_function

      ALLOWED_CHARS = /\A[\d\-\s]+\z/.freeze
      FIRST_DIGIT_WEIGHTS = [8, 7, 6, 5, 4, 3, 2].freeze
      SECOND_DIGIT_WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2].freeze
      private_constant :ALLOWED_CHARS, :FIRST_DIGIT_WEIGHTS, :SECOND_DIGIT_WEIGHTS

      def normalize(value)
        Shared.remove_non_digits(value)
      end

      def is_valid?(value)
        return false unless value.match?(ALLOWED_CHARS)

        v = normalize(value)

        return false unless v.length == 9

        base = v[0...7]
        first_digit = Mod11.mod11_check_digit(Mod11.weighted_sum(base, FIRST_DIGIT_WEIGHTS))

        return false unless first_digit == v[7].to_i

        second_digit = Mod11.mod11_check_digit(
          Mod11.weighted_sum(base + first_digit.to_s, SECOND_DIGIT_WEIGHTS)
        )

        second_digit == v[8].to_i
      end

      def format(value)
        v = normalize(value)

        return value unless v.length == 9

        "#{v[0...7]}-#{v[7...9]}"
      end
    end
  end
end
